### ISSUE TITLE
issue with adding PyTorch ROCm libs to LD_LIBRARY_PATH

### ISSUE THREAD

**Issue by ryanday36** [flux-framework/flux-core#5601]
Holger Jones just reported an issue when he tries to run with flux after putting PyTorch ROCm libs in his LD_LIBRARY_PATH:

Hey Ryan: just an fyi that I can break flux by setting LD_LIBRARY_PATH to put pytorch ROCm libs up front. 

Symptom: 
[holger@tioga11:~]$ export LD_LIBRARY_PATH=/g/g0/holger/workspace/venv/dgl_use_torch_rocm/lib64/python3.9/site-packages/torch/lib:/opt/cray/pe/papi/6.0.0.15/lib64:/opt/rocm-5.6.0/lib:/opt/rocm-5.6.0/hip/lib:/opt/rocm-5.6.0/hsa/lib:/opt/cray/pe/lib64:/opt/cray/lib64:/opt/cray/libfabric/2.1/lib64
[holger@tioga11:~]$ mxd
running: flux --parent alloc --exclusive  --nodes=1  --queue=pdebug  --time-limit=3600s  --verbose
jobid: f2VBHjtB49nw
request to allocate mask for invalid number: Invalid argument

I needed to do this to resolve a library conflict between several ML backends including tensorflow, and dgl. Torch provides its own set of RPATH'ed  ROCm libs, but they should in theory be equivalent at specific versions, say 5.6.0. Tensorflow, and dgl can use RUNPATH, so I can point them use use torch's libs - which worked fine BTW. However, the fact that it breaks flux is a snag, if not a non-starter for the AMPL/atomsci  team, since even if I unset LD_LIBRARY_PATH before an allocation, they will call flux recursively (workflows!) within an allocation, making the set/unset a bit unwieldy. 

/opt/rocm-5.x.x/lib libraries are all linked with RUNPATH; that and the fact that you may want to use Debug versions of libraries makes the use of RUNPATH appealing. However, philosophically I as a user should not be able to break flux (or any system app) using a different LD_LIBRARY_PATH.  I realize that locking down an app using RPATHs is easier said than done. So, I'm not expecting a resolution. But, I would like to get LC's perspective on this so I can make recommendations to the AMPL team and adjust expectation for ELcap ROCm ports of their machine learning applications. Maybe eventually LC can take under its wing the installation and maintenance of pytorch, tensorflow, jaxlib/xla, pyg, dgl .  I'm starting some recipe books for the installation of these via source compile but will need to hand them off eventually.

-Holger

My guess would be that he's broken hwloc somehow. I asked him to try running with mpibind off to see if the error is actually coming from mpibind.

**Comment by grondo** [comment]
Good guess, I also bet on an hwloc issue. I wonder if putting the new `LD_LIBRARY_PATH` on the `flux alloc` command line would resolve the issue for now:
```
$ flux --parent alloc --exclusive --nodes=1 --queue=pdebug --time-limit=3600s --verbose --env=LD_LIBRARY_PATH=[long path] ...
```

**Comment by grondo** [comment]
Actually, scratch that. It will probably have the same problem since the environment will be set by the job shell before the broker's are invoked.

**Comment by garlick** [comment]
This caught my eye and I didn't want to let it go by without comment

> However, philosophically I as a user should not be able to break flux (or any system app) using a different LD_LIBRARY_PATH

This does keep coming up so it would be nice to solve it.  But it is not generally permitted for a system package to subvert the dyanamic linker as explained here by redhat:

https://docs.fedoraproject.org/en-US/packaging-guidelines/#_beware_of_rpath

as a result, most system packages _can_ be broken by an errant LD_LIBRARY_PATH.  Example:
```
$ mkdir /tmp/badlib
$ touch /tmp/badlib/libc.so.6
$ LD_LIBRARY_PATH=/tmp/badlib ls
ls: error while loading shared libraries: /tmp/badlib/libc.so.6: file too short
$ 
```
Flux is packaged for redhat, so those rules apply to us.

**Comment by jonesholger** [comment]
@garlick gotcha.  Though echo'ing JohnG's RUNPATH is evil so use magic wrappers. or compile -Wl,--disable-new-dtags.

@grondo: I don't have a full reproducer I can pass to you but it's essentially these steps

ml rocm/5.6.0
ml cmake/3.23.1
ml python/3.9.12
ml rocmcc/5.6.0-magic
venv_path=$HOME/workspace/venv/dgl_use_torch_rocm
python3 -m venv $venv_path
source $venv_path/bin/activate

pip3 install --pre --no-cache-dir https://download.pytorch.org/whl/rocm5.6/torch-2.1.1%2Brocm5.6-cp39-cp39-linux_x86_64.whl

then call this script assuming it can find torch using script/torch_path.py
#!/bin/bash

BASE_DIR=`pwd`
TORCH_BASE=`$BASE_DIR/script/torch_path.py`
echo "TORCH_BASE: $BASE_DIR"
pushd $TORCH_BASE/lib
ln -s libhipblas.so libhipblas.so.1
ln -s libhipsparse.so libhipsparse.so.0
ln -s libamdhip64.so libamdhip64.so.5
ln -s librocsolver.so librocsolver.so.0
ln -s librocblas.so librocblas.so.3
ln -s librocsparse.so librocsparse.so.0
ln -s libamd_comgr.so libamd_comgr.so.2
ln -s libhsa-runtime64.so libhsa-runtime64.so.1
ln -s libdrm.so libdrm.so.2
ln -s libdrm_amdgpu.so libdrm_amdgpu.so.1
popd

torch libs just have .so extension (no version or abi) , so softlink to proper SONAME so tensorllow and dgl can pick these up in torch/lib

script/torch_path.py is
#!/usr/bin/env python3
from pathlib import Path
import sys
import os
import shutil
import torch

p = Path(torch.__file__)
sys.stdout.write(str(p.parent))

**Comment by jonesholger** [comment]
Also no luck mpibind=off 

I had in mind to source compile everything so I don't have to worry about it. But, tensorflow compile from source is a huge process for a random code team; which I have worked out.  But now I need to build xla (tensorflow jit compiler) where repo owners suggest don't clone unless you're on dev team ..  sigh!

**Comment by jonesholger** [comment]
Ok, I think Ryan's intuition is still spot on. I moved torch's libnuma .so out of the way and flux is fine.

**Comment by garlick** [comment]
I'm not sure what we could do.  BTW is this missing some small connective words or quotes?  I'm not parsing it.

> Though echo'ing JohnG's RUNPATH is evil so use magic wrappers. or compile -Wl,--disable-new-dtags.

I'm pretty sure JohnG didn't echo "RUNPATH is evil" so that interpretation is likely wrong :-)

**Comment by jonesholger** [comment]
holger@tioga29:~/workspace/dgl (use_torch_rocm *)$ more /usr/tce/packages/tce-wrapper-drivers/bin/rocmcc-wrapper | grep evil
   echo "  --enable-new-dtags       Removes, breaks our codes (and is evil)"
   echo "  -Wl,--enable-new-dtags   Removes, breaks our codes (and is evil)"
   echo "  --enable-new-dtags       Removes, breaks our codes (and evil)"
   echo "  -Wl,--enable-new-dtags   Removes, breaks our codes (and evil)"

**Comment by grondo** [comment]
>  I moved torch's libnuma .so out of the way and flux is fine.

Nice catch! libnuma is part of numactl which is probably one of hwloc's dependencies (unless it incorporates its own copy).

Library versioning is supposed to offer some protection here, but it is hard to imagine what exactly might cause this error (only hit on google is from 2011 and is a failure to call `numa_init()`)

The code in libnuma doesn't give anyone any chances to recover:
https://github.com/numactl/numactl/blob/fc999282e6663d6844773e6065bbee3270f36c19/libnuma.c#L222-L226
```c
	if (n < 1) {
		errno = EINVAL;
		numa_error("request to allocate mask for invalid number");
		exit(1);
	}
```

**Comment by jonesholger** [comment]
I think setuid/setgid should ignore LD_LIBRARY_PATH

**Comment by grondo** [comment]
Yes, that is correct `LD_LIBRARY_PATH` is ignored by for setuid/setgid binaries:

From [ld.so(8)](https://linux.die.net/man/8/ld.so)

> Using the environment variable LD_LIBRARY_PATH. Except if the executable is a set-user-ID/set-group-ID binary, in which case it is ignored.

This isn't an issue for Flux since the only setuid executable is `flux-imp` which helps the system instance launch job shells as a different user (and runs the job epilog/prolog on behalf of admins)

**Comment by grondo** [comment]
Ah, I see. I misunderstood the intent of that comment, sorry!
